### PR TITLE
[LV-234] Static patches

### PR DESCRIPTION
**Pull request by AlexandrMatveev**

### CODE REVIEW

File: Harmony/Internal/Patching/Static/AOT/AOTPatchContext.cs [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+                {
+                    var str = name;
+                    name = $"ref_{str.Remove(str.Length - 1)}";
+                }
+
+                return name;
+            }
+
+            del.Namespace = declaringType.Namespace;
+            del.Name = $"{method.ReturnType.Name}_Delegate_{methodParameters.Join(ParameterName, "_")}_{_signatureToDelegate.Count}";
```
Comment (stepan-beresnev) [inline Harmony/Internal/Patching/Static/AOT/AOTPatchContext.cs:113]: And here too, can we please make sure that all custom generated fields have some names including "harmony" so we can easily see, which methods are ours.

File: Harmony/Internal/Patching/Static/AOT/AOTPatchContext.cs [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+            {
+                var name = parameter.ParameterType.Name;
+                if (parameter.IsOut)
+                {
+                    name = $"out_{name}";
+                }
+                else if (parameter.ParameterType.IsByRef)
+                {
+                    var str = name;
+                    name = $"ref_{str.Remove(str.Length - 1)}";
```
Comment (stepan-beresnev) [inline Harmony/Internal/Patching/Static/AOT/AOTPatchContext.cs:106]: Why remove the last symbol?

**Review by stepan-beresnev — CHANGES_REQUESTED** [review]

File: Harmony/Internal/Patching/Static/AOT/DelegateBuilder.cs [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+
+        /// <summary>
+        /// Creates .ctor method
+        /// </summary>
+        private static void AddConstructor(TypeDefinition typeDefinition, TypeReference voidType, TypeReference objectType, TypeReference nativeIntType)
+        {
+            var constructor = new MethodDefinition(".ctor", ConstructorAttributes, voidType);
+            constructor.Parameters.Add(new ParameterDefinition("object", ParameterAttributes.None, objectType));
+            constructor.Parameters.Add(new ParameterDefinition("method", ParameterAttributes.None, nativeIntType));
+            constructor.ImplAttributes = MethodImplAttributes.Runtime;
```
[the remark this answers is not in the document]
Comment (AlexandrMatveev) [reply Harmony/Internal/Patching/Static/AOT/DelegateBuilder.cs:57]: image

[the remark this answers is not in the document]
Comment (AlexandrMatveev) [reply Harmony/Internal/Patching/Static/AOT/DelegateBuilder.cs:57]: It is like https://docs.microsoft.com/en-us/dotnet/api/system.runtime.compilerservices.methodimploptions?view=netframework-4.5

File: Harmony/Internal/Patching/Static/AOT/FakeMethod.cs [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+
+            var fake = new FakeMethod(
+                method.Name,
+                (MethodAttributes) method.Attributes,
+                (CallingConventions) method.CallingConvention,
+                returnType,
+                parameters,
+                declaringType);
+
+            for (var i = 0; i < parameters.Length; i++)
```
[the remark this answers is not in the document]
Comment (AlexandrMatveev) [reply Harmony/Internal/Patching/Static/AOT/FakeMethod.cs:41]: Parameter .ctor needs method instance, so i push empty parameter array in type .ctor and then fill it with parameters

**Pull request closed without merging**